engine/access/ingestion2/results_forest_test.go

```diff
@@ -0,0 +1,1688 @@
+		s.Equal(uint(receiptsPerResult), container.Size())
+		if i < 2 {
+			s.Equal(ingestion2.ResultForFinalizedBlock, container.ResultStatus())
+		} else {
+			s.Equal(ingestion2.ResultForCertifiedBlock, container.ResultStatus())
+		}
+	}
+}
+
+func (s *ResultsForestSuite) TestHasReceipt() {
+	receiptGen := s.g.ExecutionReceipts()
+
+	forest := s.createForest()
+
+	result := s.results[0]
+	receipt := receiptGen.Fixture(receiptGen.WithExecutionResult(*result))
+
+	// receipt should not be in the forest initially
+	has := forest.HasReceipt(receipt)
+	s.False(has)
+
+	_ = s.createPipeline(result)
+	added, err := forest.AddReceipt(receipt, ingestion2.ResultForCertifiedBlock)
+	s.Require().NoError(err)
+	s.True(added)
+
+	// now it should be in the forest
+	has = forest.HasReceipt(receipt)
+	s.True(has)
+}
+
+// TestOnBlockFinalized tests the OnBlockFinalized handler by creating a forest with execution and
+// consensus forks, then verifying the forks are abandoned as expected.
+//
+// This test works with 4 forks:
+// - main chain
+// - execution fork (2 results with same blocks)
+// - 2 consensus forks (different blocks and results)
+//
+// It them walks through different scenarios of finalizing the blocks to test the forest's behavior.
+func (s *ResultsForestSuite) TestOnBlockFinalized() {
+	lastSealedResult := s.results[0]
+	mainResults := s.results[1:]
+	execFork := s.executionFork(s.blocks[1:], lastSealedResult.ID())
+	consensusFork1 := s.consensusFork(2, s.blocks[0].ToHeader(), lastSealedResult.ID())
+	consensusFork2 := s.consensusFork(2, s.blocks[0].ToHeader(), lastSealedResult.ID())
+
+	// create forest and insert a sealed result
+	setupForest := func() *ingestion2.ResultsForest {
+		forest := s.createForest()
+		sealedPipeline := s.createPipeline(lastSealedResult)
+		sealedPipeline.On("SetSealed").Return().Once()
+		sealedPipeline.On("GetState").Return(optimistic_sync.StatePending).Maybe()
+		err := forest.AddSealedResult(lastSealedResult)
+		s.Require().NoError(err)
+		return forest
+	}
+
+	s.Run("marks results finalized and abandons conflicting forks", func() {
+		forest := setupForest()
+
+		_ = s.addToForest(forest, mainResults...)
+		_ = s.addToForest(forest, execFork...)
+		consensusForkPipelines := s.addToForest(forest, append(consensusFork1, consensusFork2...)...)
+
+		// finalize the second block, abandons the 2 conflicting consensus forks, and marks the first
+		// results in mainResults and execFork as finalized
+		for _, pipeline := range consensusForkPipelines {
+			pipeline.On("Abandon").Return().Once()
+		}
+
+		err := forest.OnBlockFinalized(s.blocks[1])
+		s.Require().NoError(err)
+
+		s.assertContainer(forest, mainResults[0].ID(), ingestion2.ResultForFinalizedBlock)
+		s.assertContainer(forest, mainResults[1].ID(), ingestion2.ResultForCertifiedBlock)
+		s.assertContainer(forest, execFork[0].ID(), ingestion2.ResultForFinalizedBlock)
+		s.assertContainer(forest, execFork[1].ID(), ingestion2.ResultForCertifiedBlock)
```

Comment:
Lets also check that the container status of the abandoned results was set accordingly by the forest. We check the pipeline, but I think it wouldn't hurt to also test the container directly.